bson/src/main/org/bson/codecs/pojo/CreatorExecutable.java

```diff
@@ -21,129 +21,43 @@
 import org.bson.codecs.pojo.annotations.BsonProperty;
 
 import java.lang.annotation.Annotation;
-import java.lang.reflect.Constructor;
-import java.lang.reflect.Method;
-import java.lang.reflect.Type;
-import java.util.ArrayList;
 import java.util.List;
+import java.lang.reflect.Type;
 
-import static java.lang.String.format;
-import static java.util.Arrays.asList;
+interface CreatorExecutable<T> {
+    Class<T> getType();
 
-final class CreatorExecutable<T> {
-    private final Class<T> clazz;
-    private final Constructor<T> constructor;
-    private final Method method;
-    private final List<BsonProperty> properties = new ArrayList<BsonProperty>();
-    private final Integer idPropertyIndex;
-    private final List<Class<?>> parameterTypes = new ArrayList<Class<?>>();
-    private final List<Type> parameterGenericTypes = new ArrayList<Type>();
+    List<String> getProperties();
 
-    CreatorExecutable(final Class<T> clazz, final Constructor<T> constructor) {
-        this(clazz, constructor, null);
-    }
+    Integer getIdPropertyIndex();
 
-    CreatorExecutable(final Class<T> clazz, final Method method) {
-        this(clazz, null, method);
-    }
+    List<Class<?>> getParameterTypes();
 
-    private CreatorExecutable(final Class<T> clazz, final Constructor<T> constructor, final Method method) {
-        this.clazz = clazz;
-        this.constructor = constructor;
-        this.method = method;
-        Integer idPropertyIndex = null;
+    List<Type> getParameterGenericTypes();
 
-        if (constructor != null || method != null) {
-            Class<?>[] paramTypes = constructor != null ? constructor.getParameterTypes() : method.getParameterTypes();
-            Type[] genericParamTypes = constructor != null ? constructor.getGenericParameterTypes() : method.getGenericParameterTypes();
-            parameterTypes.addAll(asList(paramTypes));
-            parameterGenericTypes.addAll(asList(genericParamTypes));
-            Annotation[][] parameterAnnotations = constructor != null ? constructor.getParameterAnnotations()
-                    : method.getParameterAnnotations();
+    T getInstance(Object... params);
 
-            for (int i = 0; i < parameterAnnotations.length; ++i) {
-                Annotation[] parameterAnnotation = parameterAnnotations[i];
+    CodecConfigurationException getError(Class<?> clazz, String msg);
 
-                for (Annotation annotation : parameterAnnotation) {
-                    if (annotation.annotationType().equals(BsonProperty.class)) {
-                        properties.add((BsonProperty) annotation);
-                        break;
-                    }
+    static Integer initializeDataFromAnnotations(Annotation[][] parameterAnnotations, List<String> properties) {
```

Comment:
A couple things strike me as a little sloppy in this method...
- Passing an empty list in, just so things can be added to it
- Doing two things at once (initializing properties list and returning the id index)

If I didn't have this method, the code would need to be duplicated in both Constructor and MethodCreatorExecutable. If I broke it into two separate functions it would require two passes over the annotations list.